osdk/src/cli.rs

```diff
@@ -476,4 +476,10 @@ pub struct CommonArgs {
         global = true
     )]
     pub encoding: Option<PayloadEncoding>,
+    #[arg(
+        long = "source-unchanged",
+        help = "Assume source code is unchanged and eagerly reuse a cached bundle if reusable",
+        global = true
+    )]
+    pub source_unchanged: bool,
```

Comment:
This looks like an unsafe workaround that would break OSDK assumptions about the source. As it may still read the configuration files.

There exist a better user interface that provides OSDK with a bundle to run `cargo osdk run /path/to/bundle`. Then the user clearly controls whether to build with the source. And it enables the user to modify non-source generated contents (such as qemu args or VM image) without rebuild.

For `cargo osdk run` that don't specify a bundle path, it should still invoke `cargo build` (and user can specify `--locked` or `offline`) and use cargo's build cache.

The Makefile can be modified that `make run` runs a built bundle instead of calling `cargo osdk run` directly.

Reply:
Thanks for the suggestion!

This option is intended to work in conjunction with `Makefile`, and users generally don't want to use it manually:

https://github.com/asterinas/asterinas/blob/0157d189ab9475ac2798ca4386d3db3c6afe0160/Makefile#L45-L48

We can remove this option, as `cargo build` already tracks dependencies and is fast enough. The option with `Makefile` saves only ~0.1s.

I use it to keep consistency with the previous code, as it doesn't track source changes. The reason why previous code never reuses cached bundles is due to two bugs (see https://github.com/asterinas/asterinas/pull/1673#issuecomment-2517283662).

Reply:
> There exist a better user interface that provides OSDK with a bundle to run cargo osdk run /path/to/bundle.

I checked the code, and if we want to run `cargo osdk run /path/to/bundle` without the `--source_unchanged` option, `cargo build` will always run and can lead to the bundle being regenerated. This is not desired and further proves that we need this option to support this feature.

Maybe the option should be renamed to something better. Also, in the Makefile, it would be better to disable `TRACK_SRC_CHANGE` by default because `cargo build` is more professional for tracking Rust dependencies.

Reply:
I think we shouldn't let OSDK build the bundle if invoked with a path to a bundle. The concept of a bundle is created for a runnable artifact that don't rely on the source to run. So it fits the need of running the already-built kernel without inspecting the source change.

There was no one using this feature and the behavior was not tested. I think it is a good attempt to fix it.

Reply:
Agreed not to build the bundle if a path is provided. But I checked the `RunArgs`, and it seems there isn't an optional positional argument? Running `cargo osdk run /path/to/bundle` simply prints the usage message.

Reply:
Yes so this is my suggested user interface. Not really asking to implement it right away. What do you think about the proposal `argo osdk run /path/to/bundle` rather than `--source-unchanged`?

Reply:
I think we currently don't have a need for `cargo osdk run /path/to/bundle`. Regarding `--source-unchanged`, it is safe to remove it.

Reply:
Ok then. Let's remove `--source-unchanged` from this PR and put it for further discussions.